[ BDH-235 ] Bundler v2 updates

adelowo opened the pull request:
- [x] Fix CTRL+C not closing processes
- [x] OTEL should use an actual config value to check if enabled
  - [x] implement OTEL
- [x] move status wait time to configuration values
- [x] Use context#Timeout as `time.After` does not garbage collect and might be hurtful for us in the long run 
- [x] Add sanity checks for http calls
- [x] Migrate Otel logic to rpc server package

adelowo committed:
fix CTRL+C not actually closing the process

adelowo committed:
fix otel collector. Set default srvice name and create new config

adelowo committed:
use config value to determine wait timeout

adelowo committed:
use context timeout cancellation as time#After does not garbage collect

adelowo committed:
remove unnneeded repetitions

adelowo committed:
create rpc middleware file and add asic test

adelowo committed:
move main to cmd package and have rpc package as an actual server

adelowo committed:
move otel logic to rpc server

adelowo committed:
add sanity checks for http calls

blewater commented on cmd/main.go:
```diff
@@ -137,17 +133,10 @@ func main() {
-	if err := b.UserMeter(otel.GetMeterProvider().Meter("bundler")); err != nil {
-		log.Fatal(err)
-	}
+	bundlerClient := createBundlerClient(mem, chain, values, eth, stdLogger)
```
This is the Bundler builder (in Ethereum sense) / backend / service / relayer and we have too many clients already :) Let's call it Service or builder.

blewater commented on cmd/main.go:
```diff
@@ -156,53 +145,44 @@ func main() {
 
-	var d = client.NewDebug(eoa, eth, mem, rep, b, chain, values.SupportedEntryPoints[0], beneficiary)
-	b.SetMaxBatch(1)
+	var d = client.NewDebug(eoa, eth, mem, rep, bundlerClient, chain, values.SupportedEntryPoints[0], beneficiary)
```
let's move the client at https://github.com/blndgs/bundler/blob/cf1a1e7a17919178744a1b02b8a9528665d0ffb3/rpc/main.go#L117 out of main also. Name this bundlerClient or erc4337Client.

blewater commented on cmd/main.go:
```diff
@@ -137,17 +133,10 @@ func main() {
-	}
+	bundlerClient := createBundlerClient(mem, chain, values, eth, stdLogger)
 
 	var check = (*checks.Standalone)(unsafe.Pointer(validator))
```
I'm concerned about line #138 I wrote because there's no compiler validation validating binary compatibility moving forward. 
I see 2 options
Make a safe type cast and panic if not successful 
or
copying the state of 'validator' to 'check'
What do you think?

adelowo replied:
fixed in https://github.com/blndgs/bundler/pull/2/commits/d235576789494003e50f914d2e1965984762e7f9

blewater commented on rpc/utils_test.go:
```diff
@@ -0,0 +1,48 @@
+	"github.com/stretchr/testify/require"
+)
+
+func Test_GetClientIPFromXFF(t *testing.T) {
```
In a test file

adelowo replied:
currently in a `utils_test.go` file

blewater commented on rpc/server.go:
```diff
@@ -0,0 +1,65 @@
+	"go.opentelemetry.io/contrib/instrumentation/github.com/gin-gonic/gin/otelgin"
+)
+
+func NewServer(values *conf.Values, logger logr.Logger, relayer *srv.Relayer,
```
maybe `NewRpcServer` because of the many clients and server components.

blewater reviewed (changes requested):
In the right direction

blewater commented:
> * [x]  Fix CTRL+C not closing processes
> * [x]  OTEL should use an actual config value to check if enabled
>   
>   * [x]  implement OTEL
> * [x]  move status wait time to configuration values
> * [x]  Use context#Timeout as `time.After` does not garbage collect and might be hurtful for us in the long run
> * [x]  Add sanity checks for http calls
> * [x]  Migrate Otel logic to rpc server package

💯

adelowo committed:
fix review and make conversion extremely safe

adelowo committed:
prevent infinite wait if tx has been dropped or does not exists

TXs can be dropped either because they are expired or the solver could not solve them. In that instance, we do not have to wait because all we do while waiting is to fetch the status of the tx on chain. If tx was dropped, then it cannot be onchain

blewater reviewed (approved):
Thank you!
